File: c/FCTest.c
#include <stdio.h>
#include <stdlib.h>

#ifndef FC_HH
#include "FileChannel.h"
#endif

#define FILENAME1 "../files/fc1.tmp"
#define FILENAME2 "../files/fc2.tmp"

FileChannel FC1 = {};
FileChannel FC2 = {};

int main() {
  FC1.file = fopen(FILENAME1, "r");

  long size = fcSize(FC1);
  long before = fcGetPosition(FC1);
  fcPosition(FC1, size / 2);
  long after = fcGetPosition(FC1);

  printf("Size of file: %-8ld\tStarting at: %-10ld\tMid point at: %ld\n",
          size, before, after);

  ByteBuffer BB = { (unsigned char *) malloc(after), size};
  fcRead(FC1, BB, size / 2);
  for (int i = 0; i < size; i++) {
    printf("%c", BB.array[i]);
  }
  fcPosition(FC1, 0);

  FC2.file = fopen(FILENAME2, "w");
  fcTransferFrom(FC2, FC1, 0, size / 2);

  fcClose(FC1);
  fcClose(FC2);

  return 0;
}

File: java/FCTest.java
import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.file.FileSystems;
import java.nio.file.StandardOpenOption;

public class FCTest {

  final static String FILENAME1 = "../files/fc1.tmp";
  final static String FILENAME2 = "../files/fc2.tmp";

  public static void main(String []args) throws IOException {
    FileChannel FC1, FC2;
    ByteBuffer BB;

    FC1 = FileChannel.open(FileSystems.getDefault()
          .getPath(FILENAME1),
          StandardOpenOption.READ);

    long size = FC1.size();
    long before = FC1.position();
    FC1.position(size / 2);
    long after = FC1.position();

    System.out.printf("Size of file: %-8d\tStarting at: %-10d\tMid point at: %d\n",
                        size, before, after);

    BB = ByteBuffer.allocate((int) size);
    BB.clear();
    FC1.read(BB, size / 2);
    for (int i = 0; i < size; i++) {
      System.out.printf("%c", BB.array()[i]);
    }
    FC1.position(0);

    FC2 = FileChannel.open(FileSystems.getDefault()
          .getPath(FILENAME2),
          StandardOpenOption.WRITE);
    FC2.transferFrom(FC1, 0, size / 2);

    FC1.close();
    FC2.close();
  }
}
